src/index/blockfilter.cpp

```diff
@@ -0,0 +1,289 @@
+
+/* The index database stores three items for each block: the encoded filter, its D256 hash, and the
+ * header. Those belonging to blocks on the active chain are indexed by height, and those belonging
+ * to blocks that have been reorganized out of the active chain are indexed by block hash. This
+ * ensures that filter data for any block that becomes part of the active chain can always be
+ * retrieved, alleviating timing concerns.
+ *
+ * Keys for the height index have the type pair<char, pair<DB_BLOCK_HEIGHT, int>>.
+ * Keys for the hash index have the type pair<char, pair<DB_BLOCK_HASH, int>>.
+ */
+constexpr char DB_FILTER = 'f';
+constexpr char DB_FILTER_HASH = 'h';
+constexpr char DB_FILTER_HEADER = 'r';
+
+constexpr char DB_BLOCK_HASH = 's';
+constexpr char DB_BLOCK_HEIGHT = 't';
+
+static std::map<BlockFilterType, BlockFilterIndex> g_filter_indexes;
+
+BlockFilterIndex::BlockFilterIndex(BlockFilterType filter_type,
+                                   size_t n_cache_size, bool f_memory, bool f_wipe)
+    : m_filter_type(filter_type)
+{
+    const std::string& filter_name = BlockFilterTypeName(filter_type);
+    if (filter_name == "") throw std::invalid_argument("unknown filter_type");
+
+    m_name = filter_name + " block filter index";
+    m_db = MakeUnique<BaseIndex::DB>(GetDataDir() / "indexes" / "blockfilter" / filter_name,
+                                     n_cache_size, f_memory, f_wipe);
+}
+
+bool BlockFilterIndex::WriteBlock(const CBlock& block, const CBlockIndex* pindex)
+{
+    CBlockUndo block_undo;
+    uint256 prev_header;
+
+    if (pindex->nHeight > 0) {
+        if (!UndoReadFromDisk(block_undo, pindex)) {
+            return false;
+        }
+
+        std::pair<uint256, uint256> read_out;
+        std::pair<char, int> prev_height_key(DB_BLOCK_HEIGHT, pindex->nHeight - 1);
+        if (!m_db->Read(std::make_pair(DB_FILTER_HEADER, prev_height_key), read_out)) {
+            return false;
+        }
+
+        uint256 expected_block_hash = pindex->pprev->GetBlockHash();
+        if (read_out.first != expected_block_hash) {
+            return error("%s: previous block header belongs to unexpected block %s; expected %s",
+                         __func__, read_out.first.ToString(), expected_block_hash.ToString());
+        }
+
+        prev_header = read_out.second;
+    }
+
+    BlockFilter filter(m_filter_type, block, block_undo);
+    std::pair<char, int> height_key(DB_BLOCK_HEIGHT, pindex->nHeight);
+
+    CDBBatch batch(*m_db);
+    batch.Write(std::make_pair(DB_FILTER, height_key),
+                std::make_pair(pindex->GetBlockHash(), filter.GetEncodedFilter()));
+    batch.Write(std::make_pair(DB_FILTER_HASH, height_key),
+                std::make_pair(pindex->GetBlockHash(), filter.GetHash()));
+    batch.Write(std::make_pair(DB_FILTER_HEADER, height_key),
+                std::make_pair(pindex->GetBlockHash(), filter.ComputeHeader(prev_header)));
+    return m_db->WriteBatch(batch);
```

Comment:
Out of scope for this PR, but is there a reason we can't write a batch of entries for more than a single block? Could we write 100 - 1000 blocks worth of entries in each batch write to speed up the migrations? This introduces some complexity, but perhaps it's worth it?

Reply:
Could be worth looking into during ThreadSync. As you note though, it's a fairly independent change.